File: src/pycropml/transpiler/antlr_py/csharp/examples/pheno_pkg/src/cs/PTQ.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class PTQ
{
    private double _tTWindowForPTQ;
    public double tTWindowForPTQ
        {
            get { return this._tTWindowForPTQ; }
            set { this._tTWindowForPTQ= value; } 
        }
    private double _kl;
    public double kl
        {
            get { return this._kl; }
            set { this._kl= value; } 
        }
    public PTQ() { }
    
    public void  CalculateModel(PhenologyState s, PhenologyState s1, PhenologyRate r, PhenologyAuxiliary a)
    {
        //- Name: PTQ -Version: 1.0, -Time step: 1
        //- Description:
    //            * Title: Phyllochron Model
    //            * Author: Pierre Martre
    //            * Reference: Modeling development phase in the 
    //                Wheat Simulation Model SiriusQuality.
    //                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    //            * Institution: INRA Montpellier
    //            * Abstract: Calculate Photothermal Quaotient 
        //- inputs:
    //            * name: tTWindowForPTQ
    //                          ** description : Thermal time window in which averages are computed
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** min : 0.0
    //                          ** max : 70000.0
    //                          ** default : 70.0
    //                          ** unit : °C d
    //                          ** uri : some url
    //                          ** inputtype : parameter
    //            * name: kl
    //                          ** description : Exctinction Coefficient
    //                          ** parametercategory : species
    //                          ** datatype : DOUBLE
    //                          ** min : 0.0
    //                          ** max : 50.0
    //                          ** default : 0.45
    //                          ** unit : 
    //                          ** uri : some url
    //                          ** inputtype : parameter
    //            * name: listTTShootWindowForPTQ_t1
    //                          ** description : List of Daily Shoot thermal time during TTWindowForPTQ thermal time period
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLELIST
    //                          ** min : 
    //                          ** max : 
    //                          ** default : [0.0]
    //                          ** unit : °C d
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: listPARTTWindowForPTQ_t1
    //                          ** description : List of Daily PAR during TTWindowForPTQ thermal time period
    //                          ** variablecategory : state
    //                          ** inputtype : variable
    //                          ** datatype : DOUBLELIST
    //                          ** min : 
    //                          ** max : 
    //                          ** default : [0.0]
    //                          ** unit : MJ m2 d
    //                          ** uri : some url
    //            * name: listGAITTWindowForPTQ
    //                          ** description : List of daily GAI over TTWindowForPTQ thermal time period
    //                          ** variablecategory : state
    //                          ** inputtype : variable
    //                          ** datatype : DOUBLELIST
    //                          ** min : 
    //                          ** max : 
    //                          ** default : [0.0, 5.2]
    //                          ** unit : m2 m-2
    //                          ** uri : some url
    //            * name: pAR
    //                          ** description : Daily Photosyntetically Active Radiation
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** default : 0.0
    //                          ** min : 0.0
    //                          ** max : 10000.0
    //                          ** unit : MJ m² d
    //                          ** uri : some url
    //                          ** inputtype : variable
    //            * name: deltaTT
    //                          ** description : daily delta TT 
    //                          ** variablecategory : auxiliary
    //                          ** inputtype : variable
    //                          ** datatype : DOUBLE
    //                          ** min : 0.0
    //                          ** max : 100.0
    //                          ** default : 0.0
    //                          ** unit : °C d
    //                          ** uri : some url
        //- outputs:
    //            * name: listPARTTWindowForPTQ
    //                          ** description :  List of Daily PAR during TTWindowForPTQ thermal time period
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLELIST
    //                          ** min : 
    //                          ** max : 
    //                          ** unit : MJ m2 d
    //            * name: listTTShootWindowForPTQ
    //                          ** description : List of Daily Shoot thermal time during TTWindowForPTQ thermal time period
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLELIST
    //                          ** min : 
    //                          ** max : 
    //                          ** unit : °C d
    //            * name: ptq
    //                          ** description : Photothermal Quotient
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** min : 0
    //                          ** max : 10000
    //                          ** unit : MJ °C-1 d m-2)
        List<double> listTTShootWindowForPTQ_t1 = s1.listTTShootWindowForPTQ;
        List<double> listPARTTWindowForPTQ_t1 = s1.listPARTTWindowForPTQ;
        List<double> listGAITTWindowForPTQ = s.listGAITTWindowForPTQ;
        double pAR = a.pAR;
        double deltaTT = a.deltaTT;
        List<double> listPARTTWindowForPTQ = new List<double>();
        List<double> listTTShootWindowForPTQ = new List<double>();
        double ptq;
        List<double> TTList = new List<double>();
        List<double> PARList = new List<double>();
        int i;
        int count;
        double SumTT;
        double parInt = 0.0d;
        double TTShoot;
        for (i=0 ; i<listTTShootWindowForPTQ_t1.Count ; i+=1)
        {
            TTList.Add(listTTShootWindowForPTQ_t1[i]);
            PARList.Add(listPARTTWindowForPTQ_t1[i]);
        }
        TTList.Add(deltaTT);
        PARList.Add(pAR);
        SumTT = TTList.Sum();
        count = 0;
        while ( SumTT > tTWindowForPTQ)
        {
            SumTT = SumTT - TTList[count];
            count = count + 1;
        }
        for (i=count ; i<TTList.Count ; i+=1)
        {
            listTTShootWindowForPTQ.Add(TTList[i]);
            listPARTTWindowForPTQ.Add(PARList[i]);
        }
        for (i=0 ; i<listTTShootWindowForPTQ.Count ; i+=1)
        {
            parInt = parInt + (listPARTTWindowForPTQ[i] * (1 - Math.Exp(-kl * listGAITTWindowForPTQ[i])));
        }
        TTShoot = listTTShootWindowForPTQ.Sum();
        ptq = parInt / TTShoot;
        s.listPARTTWindowForPTQ= listPARTTWindowForPTQ;
        s.listTTShootWindowForPTQ= listTTShootWindowForPTQ;
        s.ptq= ptq;
    }
}

File: test/Models/pheno_pkg/src/simplace/Ptq.java
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import javafx.util.*;
import java.time.LocalDateTime;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom.Element;


public class Ptq extends FWSimComponent
{
    private FWSimVariable<Double> tTWindowForPTQ;
    private FWSimVariable<Double> kl;
    private FWSimVariable<Double> pAR;
    private FWSimVariable<Double> deltaTT;
    private FWSimVariable<Double[]> listTTShootWindowForPTQ_t1;
    private FWSimVariable<Double[]> listPARTTWindowForPTQ_t1;
    private FWSimVariable<Double[]> listGAITTWindowForPTQ;
    private FWSimVariable<Double[]> listPARTTWindowForPTQ;
    private FWSimVariable<Double[]> listTTShootWindowForPTQ;
    private FWSimVariable<Double> ptq;

    public Ptq(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Ptq(){
        super();
    }
    @Override
    protected void process()
    {
        double ttTWindowForPTQ = tTWindowForPTQ.getValue();
        double tkl = kl.getValue();
        List<Double> tlistTTShootWindowForPTQ_t1 = Arrays.asList(listTTShootWindowForPTQ_t1.getValue());
        List<Double> tlistPARTTWindowForPTQ_t1 = Arrays.asList(listPARTTWindowForPTQ_t1.getValue());
        List<Double> tlistGAITTWindowForPTQ = Arrays.asList(listGAITTWindowForPTQ.getValue());
        double tpAR = pAR.getValue();
        double tdeltaTT = deltaTT.getValue();
        List<Double> tlistPARTTWindowForPTQ;
        List<Double> tlistTTShootWindowForPTQ;
        double tptq;
        List<Double> TTList = new ArrayList<>(Arrays.asList());
        List<Double> PARList = new ArrayList<>(Arrays.asList());
        int i;
        int count;
        double SumTT;
        double parInt = 0.0d;
        double TTShoot;
        for (i=0 ; i<tlistTTShootWindowForPTQ_t1.size() ; i+=1)
        {
            TTList.add(tlistTTShootWindowForPTQ_t1.get(i));
            PARList.add(tlistPARTTWindowForPTQ_t1.get(i));
        }
        TTList.add(tdeltaTT);
        PARList.add(tpAR);
        SumTT = TTList.stream().mapToDouble(Double::doubleValue).sum();
        count = 0;
        while ( SumTT > ttTWindowForPTQ)
        {
            SumTT = SumTT - TTList.get(count);
            count = count + 1;
        }
        for (i=count ; i<TTList.size() ; i+=1)
        {
            tlistTTShootWindowForPTQ.add(TTList.get(i));
            tlistPARTTWindowForPTQ.add(PARList.get(i));
        }
        for (i=0 ; i<tlistTTShootWindowForPTQ.size() ; i+=1)
        {
            parInt = parInt + (tlistPARTTWindowForPTQ.get(i) * (1 - Math.exp(-tkl * tlistGAITTWindowForPTQ.get(i))));
        }
        TTShoot = tlistTTShootWindowForPTQ.stream().mapToDouble(Double::doubleValue).sum();
        tptq = parInt / TTShoot;
        listPARTTWindowForPTQ.setValue(tlistPARTTWindowForPTQ.toArray(new Double[0]), this);
        listTTShootWindowForPTQ.setValue(tlistTTShootWindowForPTQ.toArray(new Double[0]), this);
        ptq.setValue(tptq, this);
    }

    @Override
    protected void init()
    {
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Ptq(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("ctTWindowForPTQ", "Thermal time window in which averages are computed", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 70000.0, 70.0, this));
        addVariable(FWSimVariable.createSimVariable("ckl", "Exctinction Coefficient", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 50.0, 0.45, this));
        addVariable(FWSimVariable.createSimVariable("ipAR", "Daily Photosyntetically Active Radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 10000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("ideltaTT", "daily delta TT ", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 100.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("slistTTShootWindowForPTQ", "List of Daily Shoot thermal time during TTWindowForPTQ thermal time period", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, new Double[] {0.0}, this));
        addVariable(FWSimVariable.createSimVariable("slistPARTTWindowForPTQ", "List of Daily PAR during TTWindowForPTQ thermal time period", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, new Double[] {0.0}, this));
        addVariable(FWSimVariable.createSimVariable("slistGAITTWindowForPTQ", "List of daily GAI over TTWindowForPTQ thermal time period", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, new Double[] {0.0,5.2}, this));
        addVariable(FWSimVariable.createSimVariable("sptq", "Photothermal Quotient", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 10000, null, this));

        return iFieldMap;
    }
}